### ISSUE TITLE
Odd HTTP Request behaviors

### ISSUE THREAD

**Issue by InteXX** [VBAndCs/Vazor#20]
I'm curious... do you have any idea why this works:
```
<z:if condition='ViewContext.HttpContext.Request.Path.StartsWithSegments("/Instructions")'>
  asdf
</z:if>
```
...and this doesn't?
```
<z:if condition=<%= ViewContext.HttpContext.Request.Path.StartsWithSegments("/Instructions") %>>
  asdf
</z:if>
```
The latter version produces a precompiler error for the `HttpContext` property:

> Reference to a non-shared member requires an object reference.

...and when we include a breakdown construct to separately instantiate the `ViewContext`, maybe something like this:
```
<z:if condition=<%= Me.TestContext() %>>
  asdf
</z:if>

Private Function TestContext() As String
  Dim oViewContext As ViewContext
  Dim oHttpContext As HttpContext
  Dim oRequest As HttpRequest
  Dim sResult As String

  oViewContext = New ViewContext
  oHttpContext = oViewContext.HttpContext
  oRequest = oHttpContext.Request                           ' <-- Request is empty here

  If oRequest.Path.StartsWithSegments("/Instructions") Then ' <-- NullReferenceException here
    sResult = "asdf"
  Else
    sResult = "qwer"
  End If

  Return sResult
End Function
```
Here the `HttpContext.Request` property is `Nothing`. So we get a `NullReferenceException` when we try to use `Path`. I guess that's sort of expected, given how we're creating our own `ViewContext`.

The very curious thing about it all is that the first syntax works. Somehow `ViewContext.HttpContext.Request` is available to the pipeline, from somewhere, but only when using that form.

That somewhere is the big mystery.

Your thoughts?

**Comment by VBAndCs** [comment]
I reason I've created ZML to pass such expressions to C# to be executed by the Razor engine in the proper moment. When you use this expression as an embedded XAML expression, you actually ask VB to evaluate it incautiously, before sending the response to the server, which makes sense that you will get wrong results or exceptions. If you will evaluate expressions, then you should use VBXML(Vazor) not ZML, put you will need to use ZML inside Vazor to solve such issues.
________________________________
From: Jeff Bowman ***@***.***>
Sent: Saturday, February 24, 2024 6:54 PM
To: VBAndCs/Vazor ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [VBAndCs/Vazor] Odd HTTP Request behaviors (Issue #20)


I'm curious... do you have any idea why this works:

<z:if condition='ViewContext.HttpContext.Request.Path.StartsWithSegments("/Instructions")'>
  asdf
</z:if>


...and this doesn't?

<z:if condition=<%= ViewContext.HttpContext.Request.Path.StartsWithSegments("/Instructions") %>>
  asdf
</z:if>


The latter version produces a precompiler error for the HttpContext property:

Reference to a non-shared member requires an object reference.

...and when we include a breakdown construct to separately instantiate the ViewContext, maybe something like this:

<z:if condition=<%= Me.TestContext() %>>
  asdf
</z:if>

Private Function TestContext() As String
  Dim oViewContext As ViewContext
  Dim oHttpContext As HttpContext
  Dim oRequest As HttpRequest
  Dim sResult As String

  oViewContext = New ViewContext
  oHttpContext = oViewContext.HttpContext
  oRequest = oHttpContext.Request                           ' <-- Request is empty here

  If oRequest.Path.StartsWithSegments("/Instructions") Then ' <-- NullReferenceException here
    sResult = "asdf"
  Else
    sResult = "qwer"
  End If

  Return sResult
End Function


Here the HttpContext.Request property is Nothing. So we get a NullReferenceException when we try to use Path. I guess that's sort of expected, given how we're creating our own ViewContext.

The very curious thing about it all is that the first syntax works. Somehow ViewContext.HttpContext.Request is available to the pipeline, from somewhere, but only when using that form.

That somewhere is the big mystery.

Your thoughts?

**Comment by InteXX** [comment]
I'm not sure I follow you.

This doesn't compile either:
```
<%= (Function()

  <z:if condition='ViewContext.HttpContext.Request.Path.StartsWithSegments("/Instructions")'>
    Return <a class="navbar-brand">My go-nowhere link</a>
  </z:if>

End Function).Invoke %>
```
> Multiline lambda expression is missing 'End Function'.


Is that what you're getting at when you say " ... use ZML inside Vazor?"

My goal is to access a live `Request` object within VB.NET code, rather than within a ZML expression. At this point I'm starting to wonder whether that's going to be possible. I'll poke through the ZML source when I get a chance.

**Comment by VBAndCs** [comment]
No.
You design pure ZMl pages without any VB.NET code.
Or you can use XML literals with VB code which I call vbXML
Or you can use ZML tags within vbXML.
When you use the `<%= %>` to embed expressions, VB will evaluate them before sending the page to the server, so it these expressions depend of the Razor engine or the server, you will get wrong results or exceptions. ZML solves this, because it doesn't evaluate any thing. It is just compiled to C# code, that is sent to the Razor engine to execute.
In short:  the expressions used in  `<%= %>` should only contain html tags and model properties, and avoid any objects that refers to the state after respond.

________________________________

**Comment by InteXX** [comment]
OK, got it.
